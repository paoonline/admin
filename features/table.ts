import { ITableState } from "@/models/table.type";
import { RootState } from "@/store/store";
import { createSlice } from "@reduxjs/toolkit";
import type { PayloadAction } from "@reduxjs/toolkit";

const initialState: ITableState[] = [
  {
    id: "",
    uptime: "",
    current_sales: "",
    temperature: 0,
    stock_thresholds: 0,
  },
];

export const tableSlice = createSlice({
  name: "table",
  initialState,
  reducers: {
    SET_TABLE: (_, action: PayloadAction<ITableState[]>) => {
      return action.payload
    },
  },
});

export const selectTable = (state: RootState) => state.table;

export const { SET_TABLE } = tableSlice.actions;

export default tableSlice.reducer;
