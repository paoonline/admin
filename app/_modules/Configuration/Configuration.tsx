import { Loading } from "@/app/_components/Loading/Loading";
import { selectTable } from "@/features/table";
import { ITableState } from "@/models/table.type";
import ApiClient from "@/services/api-client";
import DeleteIcon from "@mui/icons-material/Delete";
import { Box, Button, TextField } from "@mui/material";
import React, { useState } from "react";
import { useForm } from "react-hook-form";
import { useSelector } from "react-redux";
import DeleteDialog from "../DeleteDialog/DeleteDialog";

export default function Configuration({
  onEdit,
  form,
  onDelete,
}: {
  form?: ITableState;
  onEdit(data: ITableState, isError?: boolean): void;
  onDelete(data: ITableState[], isError?: boolean): void;
}) {
  const {
    register,
    formState: { errors },
    handleSubmit,
  } = useForm<ITableState>({
    defaultValues: form,
  });
  const [isLoading, setIsLoading] = useState<boolean>(false);
  const apiClient = new ApiClient();
  const tableSelector = useSelector(selectTable);

  const [openDeleteModal, setOpenDeleteModal] = useState<boolean>(false);
  const onSubmit = handleSubmit((data) => {
    setIsLoading(true);
    try {
      apiClient.SERVICES.UPDATE_DASHBOARD(data).then((res) => {
        onEdit(data);
        setIsLoading(false);
      });
    } catch (error) {
      onEdit(data, true);
      setIsLoading(false);
    }
  });

  const handleDelete = (open: boolean) => {
    handleOpenDeleteModal(true);
  };

  const handleOnDelete = () => {
    setIsLoading(true);
    try {
      apiClient.SERVICES.DELETE_DASHBOARD({ id: form?.id || "" }).then(() => {
        const newTableSelector = tableSelector.filter(
          (data) => form?.id !== data?.id
        );
        onDelete(newTableSelector);
        setIsLoading(false);
        handleDelete(false);
      });
    } catch (error) {
      onDelete([], true);
      setIsLoading(false);
      handleDelete(false);
    }
  };

  const handleOpenDeleteModal = (open: boolean) => {
    setOpenDeleteModal(open);
  };

  return (
    <form onSubmit={onSubmit}>
      <DeleteDialog
        open={openDeleteModal}
        onClose={() => handleOpenDeleteModal(false)}
        onConfirm={handleOnDelete}
      />
      {isLoading ? (
        <Loading />
      ) : (
        <React.Fragment>
          {form &&
            Object.keys(form).map((key, i) => (
              <Box
                key={key + i}
                mb={2}
                flexDirection={"row"}
                sx={{ display: "flex" }}
              >
                <Box sx={{ width: 200 }}>{key}</Box>

                <TextField
                  type={
                    key === "temperature" || key === "stock_thresholds"
                      ? "number"
                      : "text"
                  }
                  disabled={key === "id"}
                  sx={{ width: "100%" }}
                  {...register(key as keyof ITableState, { required: true })}
                  aria-invalid={
                    errors[key as keyof ITableState] ? "true" : "false"
                  }
                  error={!!errors[key as keyof ITableState]}
                />
              </Box>
            ))}
          <Box
            mt={2}
            justifyContent={"flex-end"}
            sx={{ display: "flex" }}
            gap={4}
          >
            <Button
              variant="outlined"
              onClick={() => handleDelete(true)}
              startIcon={<DeleteIcon />}
              color="warning"
            >
              Delete
            </Button>

            <Button type="submit" variant="outlined">
              Submit
            </Button>
          </Box>
        </React.Fragment>
      )}
    </form>
  );
}
