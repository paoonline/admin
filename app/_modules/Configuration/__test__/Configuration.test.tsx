import { render } from "@testing-library/react";

import { Providers } from "@/store/provider";
import Configuration from "../Configuration";
import { ITableState } from "@/models/table.type";

describe("Configuration", () => {
  it("render", () => {
    const { container } = render(
      <Configuration
        onEdit={jest.fn()}
      />,
      { wrapper: Providers }
    );
    expect(container).toMatchSnapshot();
  });
});
