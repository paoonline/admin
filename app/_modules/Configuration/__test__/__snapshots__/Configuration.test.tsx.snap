// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Configuration render 1`] = `
<div>
  <form>
    <div
      class="MuiBox-root css-1ysza1m"
    >
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedWarning MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButton-root MuiButton-outlined MuiButton-outlinedWarning MuiButton-sizeMedium MuiButton-outlinedSizeMedium css-4afo2x-MuiButtonBase-root-MuiButton-root"
        tabindex="0"
        type="button"
      >
        <span
          class="MuiButton-startIcon MuiButton-iconSizeMedium css-1d6wzja-MuiButton-startIcon"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
            data-testid="DeleteIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M6 19c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7H6zM19 4h-3.5l-1-1h-5l-1 1H5v2h14z"
            />
          </svg>
        </span>
        Delete
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium css-1rwt2y5-MuiButtonBase-root-MuiButton-root"
        tabindex="0"
        type="submit"
      >
        Submit
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
    </div>
  </form>
</div>
`;
